sponsorship_switzerland/models/completion_rules.py

```diff
@@ -236,32 +238,53 @@ def search_partner(criteria):
     #                             PRIVATE METHODS                            #
     ##########################################################################
 
-    def _generate_invoice(self, stmts_vals, st_line, partner):
-        """ Generates an invoice corresponding to the statement line read
-            in order to reconcile the corresponding move lines.
-
-        :returns dict, boolean: st_line values to update, true if invoice is
+    def _generate_invoice(self, stmts_vals, st_line, partner, ref_index):
+        """
+        Generates an invoice corresponding to the statement line read
+        in order to reconcile the corresponding move lines.
+
+        :param stmts_vals: bank statement values
+        :param st_line: current statement line values
+        :param partner: matched partner
+        :param ref_index: starting index at which the partner ref was found
+        :return: dict, boolean: st_line values to update, true if invoice is
                                 created.
         """
         # Read data in english
         res = dict()
+        ref = st_line['ref']
         product = self.with_context(lang='en_US')._find_product_id(
-            st_line['ref'])
+            partner.ref, ref)
         if not product:
             return res, False
         # Don't gengerate invoice if it's a Sponsor gift
         if product.categ_name == GIFT_CATEGORY:
             res['name'] = product.name
             contract_obj = self.env['recurring.contract'].with_context(
                 lang='en_US')
-            contract_number = int(st_line['ref'][16:21])
-            contract = contract_obj.search(
-                ['|',
-                 ('partner_id', '=', partner.id),
-                 ('correspondent_id', '=', partner.id),
-                 ('commitment_number', '=', contract_number),
-                 ('state', '!=', 'draft'),
-                 ('type', 'like', 'S')])
+            # the contract number should be found after the ref on 5 digits.
+            contract_index = ref_index + 7
+            contract_number = int(
+                ref[contract_index:contract_index+5])
+            search_criterias = [
+                '|',
+                ('partner_id', '=', partner.id),
+                ('correspondent_id', '=', partner.id),
+                ('state', '!=', 'draft'),
+                ('type', 'like', 'S')
+            ]
+            contract = contract_obj.search(search_criterias + [
+                ('commitment_number', '=', contract_number)])
+            if not contract:
+                # Maybe the ref is misaligned. Most people have a number < 10
+                # so we try to find this digit in the ref and see if we get
+                # lucky.
+                contract_number_match = re.search(
+                    partner.ref + r'0{1,4}(\d)', ref)
```

Comment:
\d can also match 0, is it on purpose?

Reply:
Yes it is, as the contract number can be 0 (unfortunately).